it/src/main/java/com/google/cloud/teleport/it/common/ResourceManagerUtils.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.it.common;
+
+import static com.google.common.hash.Hashing.goodFastHash;
+import static java.lang.Math.min;
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.hash.HashFunction;
+import com.google.re2j.Pattern;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Random;
+
+/** Common utilities for ResourceManager implementations. */
+public class ResourceManagerUtils {
+
+  public static final int MIN_PROJECT_ID_LENGTH = 4;
+  public static final int MAX_PROJECT_ID_LENGTH = 30;
+  public static final Pattern ILLEGAL_PROJECT_CHARS = Pattern.compile("[\\W_]");
+
+  private static final DateTimeFormatter TIME_FORMAT =
+      DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss-SSSSSS");
+
+  /**
+   * Generates a new id string from an existing one.
+   *
+   * @param id The id string to generate a new id from.
+   * @param targetLength The length of the new id to generate. Must be greater than 8.
+   */
+  public static String generateNewId(String id, int targetLength) {
+    if (id.length() <= targetLength) {
+      return id;
+    }
+
+    if (targetLength <= 8) {
+      throw new IllegalArgumentException("targetLength must be greater than 8");
+    }
+
+    HashFunction hashFunction = goodFastHash(32);
+    String hash = hashFunction.hashUnencodedChars(id).toString();
+    return id.substring(0, targetLength - hash.length() - 1) + "-" + hash;
+  }
+
+  /**
+   * Generates an instance id from a given string.
+   *
+   * @param baseString The string to generate the id from.
+   * @return The instance id string.
+   */
+  public static String generateInstanceId(
```

Comment:
What's the purpose of this method? Is it supposed to generate Spanner instance IDs or an arbitrary resource ID? Please add this to javadoc. If Spanner IDs, then it should be moved to `SpannerResourceManagerUtils`. If any resource ID, then I'd change its name to `generateResourceId` (but then how is it different from `generateNewId`? please rename and/or add javadoc details), and remove comments like "so it doesn't violate spanner's instance".

Reply:
It was intended to generate Instance ID's for both BigTable and Spanner as they both have a concept of Instances with the same constraints. However, I agree having a `generateResourceId` function is cleaner and more adaptable. I refactored to support this.